Add BooleanFilter and NumericFilter

Simperfit opened the pull request:
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | no
| Fixed tickets | #352 
| License       | MIT
| Doc PR        | api-platform/doc#38

I have added an boolean filter based on the order filter with the test.

- [x] [Boolean Filter]
- [x] [Numeric Filter]

Simperfit commented:
@dunglas There are some behat errors on my local : 
`   And the response should be equal to                                 # Sanpi\Behatch\Context\RestContext::theResponseShouldBeEqualTo()
      """
      <?xml version="1.0"?>
      <response><id>1</id><name>XML!</name><alias/><description/><dummyDate/><dummyPrice/><jsonData/><relatedDummy/><dummy/><relatedDummies/><nameConverted/></response>
      """
      The string '<?xml version="1.0"?>
      <response><id>1</id><name>XML!</name><alias/><description/><dummyDate/><dummyPrice/><jsonData/><relatedDummy/><dummy/><relatedDummies/><nameConverted/></response>' is not equal to the response of the current page
      
      +--[ HTTP/1.1 201 | http://example.com/dummies | KernelDriver ]
      |
      |  <?xml version="1.0"?>
      |  <response><id>1</id><name>XML!</name><alias/><description/><dummyDate/><dummyPrice/><jsonData/><relatedDummy/><dummyBoolean/><dummy/><relatedDummies/><nameConverted/></response>
      |
`

`  And the JSON should be equal to:      
                                 # Sanpi\Behatch\Context\JsonContext::theJsonShouldBeEqualTo()
      """
      {
        "@context": "/contexts/Dummy",
....
      The json is equal to:
      {
          "@context": "\/contexts\/Dummy"
.....` 
Why does the app return things like :  "@context": "\/contexts\/Dummy", is it something linked to what i'm doing ?

gorghoa commented on features/doctrine/boolean-filter.feature:
```diff
@@ -0,0 +1,222 @@
+Feature: Boolean filter on collections
+  In order to retrieve ordered large collections of resources
+  As a client software developer
+  I need to retrieve collections with boolean balue
```
balue => value

dunglas commented:
> Why does the app return things like : "@context": "\/contexts\/Dummy",is it something linked to what i'm doing ?

No it's normal, `\` can be safely ignored. The problem is probably in the body of the JSON document.

dunglas commented on src/Bridge/Doctrine/Orm/Filter/BooleanFilter.php:
```diff
@@ -0,0 +1,136 @@
+     */
+    private $booleanParameterName;
+
+    /**
```
Useless docblock

dunglas commented on src/Bridge/Doctrine/Orm/Filter/BooleanFilter.php:
```diff
@@ -0,0 +1,136 @@
+            }
+
+            $boolean = strtoupper($boolean);
+         if (!in_array($boolean, ['0', '1','TRUE','FALSE'])) {
```
Maybe can you rely on `filter_var()` instead of testing boolean values manually? It will also add support for `on`/`off`, lower cases versions and so on.

Simperfit replied:
What about empty values ? since filter_var return false on empty strings, do we have to check them with the filter_var or is it ok to return boolean false on empty strings ?

dunglas replied:
You're right, it's probably better to exclude empty values (but use `filter_var` for others).

Are you ok with that @api-platform/core-team?

Simperfit replied:
I've added an test on empty value, tell me if its ok

dunglas commented on src/Bridge/Doctrine/Orm/Filter/BooleanFilter.php:
```diff
@@ -0,0 +1,136 @@
+
+            $description[sprintf('%s[%s]', $this->booleanParameterName, $property)] = [
+                'property' => $property,
+                'type' => 'string',
```
`boolean`?

Simperfit replied:
It is ok to filter only on boolean like in the DateFilter or do we have to include all type of property ?

dunglas commented on src/Bridge/Doctrine/Orm/Filter/BooleanFilter.php:
```diff
@@ -0,0 +1,136 @@
+    /**
+     * @var string Keyword used to retrieve the value.
+     */
+    private $booleanParameterName;
```
IMO it would be cleaner to support the same syntax than for the search filter:

http://example.com/users?enabled=true looks better to me than http://example.com/users?boolean[users]=true

WDYT?

By the way, it will simplify this PR (no more configuration required).

Simperfit replied:
Is in your exemple 'enabled' the field you want to check ?

I think it's better than what I proposed, it will be better to understand and to use.

Ill work on this & modify the test.

Thank you for the help,

dunglas replied:
> Is in your exemple 'enabled' the field you want to check ?

Yes, exactly.

theofidry replied:
I also prefer `http://example.com/users?enabled=true`, the case where you want to search for the *string* true instead of the boolean value is such an edge case that I would not bother with it.

Simperfit replied:
@theofidry  what about the empty strings ?

theofidry replied:
on a boolean I would expect `null` :)

sstok commented:
FYI it's `[x] Boolean Filter` not `[Boolean Filter] - X` :smile: that's how you get those checkmark boxes.

Simperfit commented:
@sstok Thank for the information, i didn't know that

Simperfit commented:
@dunglas Could you help me to determine  the scope of the `NumericFilter`?
Is it to check a field's type or/and to work on it, like the `DateFilter`  </>/=/between  does ?
Thanks you :)

Simperfit commented:
Added numeric Filter

dunglas commented:
@Simperfit Can you try to fix Travis tests?

Simperfit commented:
I've modified the test with the added booleanValue.
Travis is green :)

Simperfit commented:
There is something I don't understand btw,

the` features/json-ld/context.feature:6` test does not pass on local neither it does on appveyor, is it related to the filters i've added ? the dummy is not on the same line on travis & on local.

dupuchba commented on features/bootstrap/FeatureContext.php:
```diff
@@ -215,6 +215,25 @@ public function thereIsDummyObjectsWithDummyPrice($nb)
 
         $this->manager->flush();
     }
+    /**
```
New line plz

Simperfit replied:
added :)

teohhanhui commented:
@Simperfit That's #454

theofidry commented on features/bootstrap/HydraContext.php:
```diff
@@ -180,9 +180,9 @@ public function assertPropertyIsRequired($propertyName, $className)
-     *
      * @throws Exception
+     *
+     * @return array
```
Indeed it's better to put `@throws` tag before the `@return`, but we should check that on the whole provide and fix it everywhere. It would be better to do so in another PR as well to avoid polluting this one :)

teohhanhui replied:
See #452

This needs rebasing.

theofidry commented on src/Action/PostCollectionAction.php:
```diff
@@ -42,8 +42,9 @@ public function __construct(SerializerInterface $serializer)
-        list($resourceClass, $operationName, , $format) = $this->extractAttributes($request);
-        $context = ['resource_class' => $resourceClass, 'collection_operation_name' => $operationName];
+
+        list($resourceClass, $operationName,$itemOperation, $format) = $this->extractAttributes($request);
```
why this change?

Simperfit replied:
It does trigger an syntax error with php7.0.4

theofidry replied:
We should fix that in another PR then, it's not related to your filter and is a different issue :)

teohhanhui replied:
> It does trigger an syntax error with php7.0.4

How is that possible? Travis tests are using latest PHP 7, which is 7.0.4, and the tests are passing.

theofidry commented on src/Bridge/Doctrine/Orm/Filter/BooleanFilter.php:
```diff
@@ -0,0 +1,145 @@
+class BooleanFilter extends AbstractFilter
+{
+    /**
+     * @var RequestStack
```
unnecessary phpdoc block

Simperfit replied:
which one ?

theofidry commented on src/Bridge/Doctrine/Orm/Filter/NumericFilter.php:
```diff
@@ -0,0 +1,159 @@
+    ];
+
+    /**
+     * @var RequestStack
```
unnecessary phpdoc block

theofidry commented on tests/Bridge/Doctrine/Orm/Filter/BooleanFilterTest.php:
```diff
@@ -0,0 +1,246 @@
+     */
+    public function filterProvider()
+    {
+        return [
```
we can use `yield` here to make it more readable :)

```php
// without yield
return [
  [arg1, arg2, ...],
  [arg1, arg2, ...],
];

// with yield
yield [arg1, arg2, ...];
yield [arg1, arg2, ...];
```

Simperfit replied:
Do I have to modify the other tests ?

teohhanhui replied:
I don't see how using a generator is more readable. In fact, it's potentially more puzzling, and PHPUnit documentation does not mention support for generators as data providers.

teohhanhui replied:
If we're really concerned about readability, we'd do well to add string keys.

```php
return [
    'true value' => [...],
    'false value' => [...],
    'non-boolean value' => [...],
    '0 value' => [...],
    '1 value' => [...],
];
```

theofidry replied:
But are the keys passed? When using a data provider, you get the content of the array values, not the keys. Or did I miss something?

theofidry replied:
@Simperfit could be done but not in this PR in any case

teohhanhui replied:
The keys are used by PHPUnit for display when the test case fails (as a label to help us identify the exact test case).

teohhanhui replied:
https://phpunit.de/manual/current/en/writing-tests-for-phpunit.html#writing-tests-for-phpunit.data-providers

See example 2.6

theofidry replied:
nice, then indeed it would be better to put the current comments as keys

teohhanhui replied:
Please add a test case with nested property.

Simperfit replied:
added one

Simperfit commented:
ping @api-platform/core-team.

Test pass on local, both filters are working with the Standard edition.

teohhanhui commented on features/invalid_data.feature:
```diff
@@ -23,6 +23,7 @@ Feature: Handle properly invalid data submitted to the API
       "@type": "Dummy",
       "description": null,
       "dummy": null,
+      "dummyBoolean":null,
```
Missing space

teohhanhui commented on features/json-ld/context.feature:
```diff
@@ -20,11 +20,12 @@ Feature: JSON-LD contexts generation
-      "user": "/users"
+      "user": "/users",
+      "dummy": "/dummies"
+
```
Extra blank line

teohhanhui commented on features/json-ld/context.feature:
```diff
@@ -20,11 +20,12 @@ Feature: JSON-LD contexts generation
       "customIdentifierDummy": "/custom_identifier_dummies",
       "customNormalizedDummy": "/custom_normalized_dummies",
       "customWritableIdentifierDummy": "/custom_writable_identifier_dummies",
-      "dummy": "/dummies",
```
Why the change?

Simperfit replied:
I think it was because of  #454

teohhanhui commented on features/relation.feature:
```diff
@@ -70,6 +70,7 @@ Feature: Relations support
       "@type": "Dummy",
       "description": null,
       "dummy": null,
+      "dummyBoolean":null,
```
Missing space

teohhanhui commented on features/json-ld/context.feature:
```diff
@@ -41,6 +42,7 @@ Feature: JSON-LD contexts generation
               "hydra": "http://www.w3.org/ns/hydra/core#",
               "description": "https://schema.org/description",
               "dummy": "#Dummy/dummy",
+              "dummyBoolean":"#Dummy/dummyBoolean",
```
Missing space

teohhanhui commented on features/unknown_attributes.feature:
```diff
@@ -24,6 +24,7 @@ Feature: Ignore unknown attributes
       "@type": "Dummy",
       "description": null,
       "dummy": null,
+      "dummyBoolean":null,
```
Missing space

teohhanhui commented on src/Bridge/Doctrine/Orm/Filter/BooleanFilter.php:
```diff
@@ -0,0 +1,143 @@
+/**
+ * Filter by a boolean value.
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
```
You should only put yourself as the author here.

teohhanhui commented on src/Bridge/Doctrine/Orm/Filter/BooleanFilter.php:
```diff
@@ -0,0 +1,143 @@
+
+                $field = $propertyParts['field'];
+            }
+            $valueParameter = QueryNameGenerator::generateParameterName(sprintf('%s_%s', $field, $boolean));
```
`QueryNameGenerator::generateParameterName($field)` is enough.

teohhanhui commented on src/Bridge/Doctrine/Orm/Filter/BooleanFilter.php:
```diff
@@ -0,0 +1,143 @@
+
+            $queryBuilder
+                ->andWhere(sprintf('%s.%s = :%s', $alias, $field, $valueParameter))
+                ->setParameter($valueParameter, filter_var($boolean, FILTER_VALIDATE_BOOLEAN, FILTER_NULL_ON_FAILURE));
```
Let's do the `filter_var` only once above.

teohhanhui commented on src/Bridge/Doctrine/Orm/Filter/BooleanFilter.php:
```diff
@@ -0,0 +1,143 @@
+    /**
+     * {@inheritdoc}
+     */
+    protected function extractProperties(Request $request) : array
```
`extractProperties` is already in `AbstractFilter`, and it also handles fixing of nested properties.

teohhanhui commented on src/Bridge/Doctrine/Orm/Filter/BooleanFilter.php:
```diff
@@ -0,0 +1,143 @@
+        $propertyParts = $this->splitPropertyParts($property);
+        $metadata = $this->getNestedMetadata($resourceClass, $propertyParts['associations']);
+
+        return 'boolean' == $metadata->getTypeOfField($propertyParts['field']) ? true : false;
```
```php
use Doctrine\DBAL\Types\Type as DBALType;
...
    return DBALType::BOOLEAN === $metadata->getTypeOfField($propertyParts['field']);
```

teohhanhui commented on src/Bridge/Doctrine/Orm/Filter/NumericFilter.php:
```diff
@@ -0,0 +1,157 @@
+class NumericFilter extends AbstractFilter
+{
+    /*
+     * Type of numeric in Doctrine see http://doctrine-orm.readthedocs.org/projects/doctrine-dbal/en/latest/reference/types.html
```
Use proper PHPDoc:

```
@see http://...
```

And remove the extra blank line.

teohhanhui commented on src/Bridge/Doctrine/Orm/Filter/NumericFilter.php:
```diff
@@ -0,0 +1,157 @@
+    /**
+     * {@inheritdoc}
+     */
+    protected function extractProperties(Request $request) : array
```
Same as above. Please remove.

teohhanhui commented on src/Bridge/Doctrine/Orm/Filter/NumericFilter.php:
```diff
@@ -0,0 +1,157 @@
+     *
+     */
+    const DOCTRINE_NUMERIC_TYPES = [
+        'bigint' => true,
```
I'd prefer using the class constants:

```php
// use Doctrine\DBAL\Types\Type as DBALType;

const DOCTRINE_NUMERIC_TYPES = [
    DBALType::BIGINT => true,
    DBALType::DECIMAL => true,
    DBALType::FLOAT => true,
    DBALType::INTEGER => true,
    DBALType::SMALLINT => true,
];
```

sorted, and `time` is not a numeric type, according to the linked Doctrine reference.

teohhanhui commented on src/Bridge/Doctrine/Orm/Filter/NumericFilter.php:
```diff
@@ -0,0 +1,157 @@
+/**
+ * Filter by a numeric value.
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
```
Same here. You're the only author for this class.

teohhanhui commented on src/Bridge/Doctrine/Orm/Filter/NumericFilter.php:
```diff
@@ -0,0 +1,157 @@
+
+                $field = $propertyParts['field'];
+            }
+            $valueParameter = QueryNameGenerator::generateParameterName(sprintf('%s_%s', $field, 'equals'));
```
`QueryNameGenerator::generateParameterName($field)` is enough.

teohhanhui commented on src/Bridge/Symfony/Bundle/Resources/config/doctrine_orm.xml:
```diff
@@ -46,12 +46,20 @@
             <argument type="service" id="doctrine" />
             <argument type="service" id="request_stack" />
         </service>
-
```
Accidental removal of blank line.

teohhanhui commented on src/Bridge/Symfony/Bundle/Resources/config/doctrine_orm.xml:
```diff
@@ -46,12 +46,20 @@
+        <service id="api_platform.doctrine.orm.numeric_filter" class="ApiPlatform\Core\Bridge\Doctrine\Orm\Filter\NumericFilter" public="false" abstract="true">
+            <argument type="service" id="doctrine" />
+            <argument type="service" id="request_stack" />
+        </service>
```
There should be a blank line after this section.

Simperfit replied:
i've added one

teohhanhui commented on tests/Bridge/Doctrine/Orm/Filter/BooleanFilterTest.php:
```diff
@@ -0,0 +1,242 @@
+use Symfony\Component\HttpFoundation\RequestStack;
+
+/**
+ * @author Théo FIDRY <theo.fidry@gmail.com>
```
You're the only author.

teohhanhui commented on tests/Bridge/Doctrine/Orm/Filter/NumericFilterTest.php:
```diff
@@ -0,0 +1,201 @@
+     */
+    public function filterProvider()
+    {
+        return [
```
Same here. Please add a test case for nested property.

Simperfit replied:
i've added  one

teohhanhui commented on tests/Fixtures/TestBundle/Entity/Dummy.php:
```diff
@@ -74,6 +74,13 @@ class Dummy
     /**
+     * @var bool A dummy boolean.
+     *
+     * @ORM\Column(type="boolean",nullable=true)
```
Missing space after comma.

teohhanhui commented on src/Bridge/Doctrine/Orm/Filter/NumericFilter.php:
```diff
@@ -0,0 +1,148 @@
+    }
+
+    /**
+     * Determines whether the given property is a numerical or not.
```
Let's stick to the same term "numeric".

Simperfit replied:
We can change that to numerical in every method if you think it's better way to understand what it does.

teohhanhui replied:
What I'm saying is that we should stick to "numeric", because that's what PHP uses (e.g. `is_numeric`).

Simperfit replied:
My bad, I didn't undertstand that way ;), will change that.

teohhanhui commented on src/Bridge/Doctrine/Orm/Filter/NumericFilter.php:
```diff
@@ -0,0 +1,148 @@
+     *
+     * @return bool
+     */
+    private function isNumericalField(string $property, string $resourceClass) : bool
```
`isNumericField`

teohhanhui commented on tests/Bridge/Doctrine/Orm/Filter/BooleanFilterTest.php:
```diff
@@ -0,0 +1,250 @@
+                ],
+                [
+                   'dummyBoolean' => '1',
+                   'name' => '1',
```
Why are we testing the `BooleanFilter` with non-boolean fields (e.g. `name`), or is this intentional?

Simperfit replied:
This is intentional to see what it does using wrong type values.

But as I can see, it does use that value, need to check why.

teohhanhui commented on tests/Bridge/Doctrine/Orm/Filter/BooleanFilterTest.php:
```diff
@@ -0,0 +1,250 @@
+                    'properties' => ['id' => null, 'name' => null, 'relatedDummy.id' => null],
+                ],
+                [
+                    'relatedDummy.id' => '1',
```
This does not show that it works with nested properties, because the field type is wrong so it's just being skipped.

teohhanhui commented on tests/Bridge/Doctrine/Orm/Filter/NumericFilterTest.php:
```diff
@@ -0,0 +1,209 @@
+                    'properties' => ['id' => null, 'name' => null, 'dummyBoolean' => null],
+                ],
+                [
+                    'id' => 21,
```
I'd rather we don't use the `id` for testing the `NumericFilter`.

Simperfit replied:
Do I have to add multiple numeric type field to test each doctrine type ?

teohhanhui replied:
Just use `dummyPrice` :smile:

teohhanhui commented on tests/Bridge/Doctrine/Orm/Filter/NumericFilterTest.php:
```diff
@@ -0,0 +1,209 @@
+    public function filterProvider()
+    {
+        return [
+            // test with true value
```
What do you mean by "true" value?

Simperfit replied:
The boolean `true` value

teohhanhui replied:
But this is not the `BooleanFilter`. I think your comments here need to be updated.

teohhanhui commented on tests/Bridge/Doctrine/Orm/Filter/NumericFilterTest.php:
```diff
@@ -0,0 +1,209 @@
+                ],
+                sprintf('SELECT o FROM %s o where o.id = :id_123456abcdefg', Dummy::class),
+            ],
+            // test with non-boolean value
```
non-numeric

teohhanhui commented on tests/Bridge/Doctrine/Orm/Filter/NumericFilterTest.php:
```diff
@@ -0,0 +1,209 @@
+                ],
+                sprintf('SELECT o FROM %s o left join o.relateddummy relateddummy_123456abcdefg where relateddummy_123456abcdefg.id = :id_123456abcdefg', Dummy::class),
+            ],
+            // test with multiple 1 value
```
Multiple mismatched comments for the test data set.

teohhanhui commented on tests/Fixtures/TestBundle/Entity/Dummy.php:
```diff
@@ -22,7 +22,7 @@
  * @author Kévin Dunglas <dunglas@gmail.com>
  *
- * @Resource(attributes={"filters"={"my_dummy.search", "my_dummy.order", "my_dummy.date", "my_dummy.range"}})
+ * @Resource(attributes={"filters"={"my_dummy.search", "my_dummy.order", "my_dummy.date", "my_dummy.range","my_dummy.boolean","my_dummy.numeric"}})
```
Missing space after comma.

teohhanhui commented on src/Bridge/Doctrine/Orm/Filter/NumericFilter.php:
```diff
@@ -0,0 +1,148 @@
+use ApiPlatform\Core\Bridge\Doctrine\Orm\Util\QueryNameGenerator;
+use Doctrine\Common\Persistence\ManagerRegistry;
+use Doctrine\ORM\QueryBuilder;
+use Doctrine\DBAL\Types\Type as DBALType;
```
Please check StyleCI results. We enforce ordered use statements.

Simperfit replied:
I didn't push my last commit, will do that tonight

Simperfit commented:
@teohhanhui I've modified the tests,

I've added one question about the term, I could refactor the class / test and implementation if it's better with numerical.

Thank you for the time and for the revision of this PR :+1:

teohhanhui commented:
There are still a few issues which have not been addressed, but otherwise it looks good.

Simperfit commented:
@teohhanhui done fixing the issues i've seen, thanks :).

theofidry commented:
@teohhanhui LGTM, test are failing due to random order, related to #454.

Simperfit commented:
i've rebsed with master

dunglas commented:
Thank you @Simperfit for this amazing work!

Simperfit commented:
NP @dunglas

Thanks to everyone who helped me ;).
